File: python/jskit/jsmake/constraints.py
def compileConstraintEquality(equality, refs):
    """
    >>> refs = dict(self="/Root", one='/One', two='/Two', n=10)
    >>> sorted(compileConstraintEquality("self.width = 12", refs).items())
    [('constant', 12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Root'), ('relation', '$UILayoutRelation.equal')]
    >>> sorted(compileConstraintEquality("self.height = n", refs).items())
    [('constant', 10), ('firstAttribute', '$UILayoutAttribute.height'), ('firstItem', '/Root'), ('relation', '$UILayoutRelation.equal')]
    >>> sorted(compileConstraintEquality("one.top = self.top", refs).items())
    [('firstAttribute', '$UILayoutAttribute.top'), ('firstItem', '/One'), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.top'), ('secondItem', '/Root')]
    >>> sorted(compileConstraintEquality("one.top = self.top + 12", refs).items())
    [('constant', 12), ('firstAttribute', '$UILayoutAttribute.top'), ('firstItem', '/One'), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.top'), ('secondItem', '/Root')]
    >>> sorted(compileConstraintEquality("one.top = self.top + n", refs).items())
    [('constant', 10), ('firstAttribute', '$UILayoutAttribute.top'), ('firstItem', '/One'), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.top'), ('secondItem', '/Root')]
    >>> sorted(compileConstraintEquality("one.top = self.top - 12", refs).items())
    [('constant', -12), ('firstAttribute', '$UILayoutAttribute.top'), ('firstItem', '/One'), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.top'), ('secondItem', '/Root')]
    >>> sorted(compileConstraintEquality("one.top = self.top - n", refs).items())
    [('constant', -10), ('firstAttribute', '$UILayoutAttribute.top'), ('firstItem', '/One'), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.top'), ('secondItem', '/Root')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2", refs).items())
    [('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2.5", refs).items())
    [('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2.5), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2 + 12", refs).items())
    [('constant', 12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2.5 + 12", refs).items())
    [('constant', 12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2.5), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2 + n", refs).items())
    [('constant', 10), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2.5 + n", refs).items())
    [('constant', 10), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2.5), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2 - 12", refs).items())
    [('constant', -12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2.5 - 12", refs).items())
    [('constant', -12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2.5), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2 - n", refs).items())
    [('constant', -10), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("two.width = one.width * 2.5 - n", refs).items())
    [('constant', -10), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Two'), ('multiplier', 2.5), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("self.width = 12 @ 100", refs).items())
    [('constant', 12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Root'), ('priority', 100), ('relation', '$UILayoutRelation.equal')]
    >>> sorted(compileConstraintEquality("self.width = 12 @ $UILayoutPriority.defaultLow", refs).items())
    [('constant', 12), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Root'), ('priority', '$UILayoutPriority.defaultLow'), ('relation', '$UILayoutRelation.equal')]
    >>> sorted(compileConstraintEquality("self.width = one.width * 2 + 3 @ 100", refs).items())
    [('constant', 3), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Root'), ('multiplier', 2), ('priority', 100), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    >>> sorted(compileConstraintEquality("self.width = one.width * 2 + 3 @ $UILayoutPriority.defaultLow", refs).items())
    [('constant', 3), ('firstAttribute', '$UILayoutAttribute.width'), ('firstItem', '/Root'), ('multiplier', 2), ('priority', '$UILayoutPriority.defaultLow'), ('relation', '$UILayoutRelation.equal'), ('secondAttribute', '$UILayoutAttribute.width'), ('secondItem', '/One')]
    """
    def resolveNumber(token):
        if token in refs:
            return refs[token]
        if token[0:1] == '$':
            return token
        if '.' in token:
            return float(token);
        return int(token)
    equality = equality.replace(' ', '')
    l = len(equality)
    if l == 0:
        raise Exception(u"Empty equality found")
    i = 0
    constraint = dict()
    item1Name = ''
    while i < l and equality[i] != '.':
        item1Name += equality[i]
        i += 1
    if len(item1Name) == 0:
        raise Exception("Expecting item name to start equality: %s" % equality)
    if i == l:
        raise Exception("Expecting '.' in equality: %s" % equality)
    if item1Name not in refs:
        raise Exception("Ref name '%s' not found in equality: %s", (item1Name, equality))
    constraint['firstItem'] = refs[item1Name]
    i += 1
    item1Attr = ''
    while i < l and equality[i] not in ('<', '>', '='):
        item1Attr += equality[i]
        i += 1
    if len(item1Attr) == 0:
        raise Exception("Expecting item attribute in equality: %s" % equality)
    if i == l:
        raise Exception("Expecting '=', '<=', or '>=' in equality: %s" % equality)
    if item1Attr not in ('left', 'right', 'top', 'bottom', 'leading', 'trailing', 'width', 'height', 'centerX', 'centerY'):
        raise Exception("Attribute name '%s' not valid in equality: %s", (item1Attr, equality))
    constraint['firstAttribute'] = '$UILayoutAttribute.%s' % item1Attr
    relation = '$UILayoutRelation.equal'
    if equality[i] in ('<', '>'):
        i += 1
        if i == l or equality[i] != '=':
            raise Exception("Expecting '=' after %s in equality: %s" % (equality[i - 1], equality))
        if equality[i - 1] == '<':
            relation = '$UILayoutRelation.lessThanOrEqual'
        else:
            relation = '$UILayoutRelation.greaterThanOrEqual'
    i += 1
    if i == l:
        raise Exception("Expecting right hand side of equality: %s", equality)
    constraint['relation'] = relation
    token = ''
    while i < l and equality[i] not in ('.', '*', '+', '-', '@'):
        token += equality[i]
        i += 1
    if len(token) == 0:
        raise Exception("Expecting item or constant on right of equality: %s", equality)
    if i == l:
        constraint['constant'] = resolveNumber(token)
    else:
        if equality[i] == '.':
            i += 1
            item2Name = token
            if item2Name not in refs:
                raise Exception("Ref name '%s' not found in equality: %s", (item2Name, equality))
            constraint['secondItem'] = refs[item2Name]
            item2Attr = ''
            while i < l and equality[i] not in ('*', '+', '-', '@'):
                item2Attr += equality[i]
                i += 1
            if len(item2Attr) == 0:
                raise Exception("Expecting item attribute in equality: %s" % equality)
            if item2Attr not in ('left', 'right', 'top', 'bottom', 'leading', 'trailing', 'width', 'height', 'centerX', 'centerY'):
                raise Exception("Attribute name '%s' not valid in equality: %s", (item2Attr, equality))
            constraint['secondAttribute'] = '$UILayoutAttribute.%s' % item2Attr
        elif equality[i] == '@':
            constraint['constant'] = resolveNumber(token)
        else:
            raise Exception("Unexpected token in equality at %d: %s" % (i, equality))
        if i < l and equality[i] == '*':
            i += 1
            token = ''
            while i < l and equality[i] not in ('+', '-', '@'):
                token += equality[i]
                i += 1
            if len(token) == 0:
                raise Exception("Expecting multiplier in equality: %s" % equality)
            constraint['multiplier'] = resolveNumber(token)
        if i < l and equality[i] in ('-', '+'):
            sign = -1 if equality[i] == '-' else 1
            i += 1
            token = ''
            while i < l and equality[i] not in ('@',):
                token += equality[i]
                i += 1
            if len(token) == 0:
                raise Exception("Expecting constraint in equality: %s" % equality)
            constraint['constant'] = sign * resolveNumber(token)
        if i < l and equality[i] == '@':
            i += 1
            token = equality[i:]
            if len(token) == 0:
                raise Exception("Expecting priority in equality: %s" % equality)
            constraint['priority'] = resolveNumber(token)
    return constraint

File: Command/jskit/Constraints.js
// Copyright 2020 Breakside Inc.
//
// Licensed under the Breakside Public License, Version 1.0 (the "License");
// you may not use this file except in compliance with the License.
// If a copy of the License was not distributed with this file, you may
// obtain a copy at
//
//     http://breakside.io/licenses/LICENSE-1.0.txt
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// #import Foundation
'use strict';

JSGlobalObject.constraintsFromSpecShorthand = function(shorthand){
    var constraints = [];
    var references = JSCopy(shorthand.references);
    if (shorthand.equalities){
        for (let i = 0, l = shorthand.equalities.length; i < l; ++i){
            let constraint = constraintFromEquality(shorthand.equalities[i], references);
            constraints.push(constraint);
        }
    }
    return constraints;
};

JSGlobalObject.constraintFromEquality = function(equality, references){
    var resolveNumber = function(token){
        if (token in references){
            return references[token];
        }
        if (token.match(/^\d+$/)){
            return parseInt(token);
        }
        if (token.match(/^\d+\.\d+$/)){
            return parseFloat(token);
        }
        return token;
    };
    var eq = equality.replace(/ /g, '');
    var l = eq.length;
    if (l === 0){
        throw new Error("Empty equality found");
    }
    var i = 0;
    var constraint = {};
    var token = '';
    while (i < l && eq[i] !== '.'){
        token += eq[i];
        ++i;
    }
    if (token.length === 0){
        throw new Error("Expecting item name to start equality: %s".sprintf(equality));
    }
    if (i === l){
        throw new Error("Expecting '.' in equality: %s".sprintf(equality));   
    }
    if (!(token in references)){
        constraint.firstItemName = token;
    }else{
        constraint.firstItem = references[token];
    }
    i += 1;
    token = '';
    while (i < l && eq[i] !== '<' && eq[i] !== '>' && eq[i] !== '=' && eq[i] !== '.'){
        token += eq[i];
        ++i;
    }
    if (token.length === 0){
        throw new Error("Expecting item attribute in start equality: %s".sprintf(equality));
    }
    if (i === l){
        throw new Error("Expecting '=' or '<=' or '>=' in equality: %s".sprintf(equality));   
    }
    if (eq[i] == '.'){
        if (!allowedProperties.has(token)){
            if (!allowedAttributes.has(token)){
                throw new Error("Property name '%s' not valid in equality: %s".sprintf(token, equality));
            }
        }
        constraint.firstProperty = token;
        i += 1;
        token = '';
        while (i < l && eq[i] !== '<' && eq[i] !== '>' && eq[i] !== '=' && eq[i] !== '.'){
            token += eq[i];
            ++i;
        }
        if (token.length === 0){
            throw new Error("Expecting item attribute in start equality: %s".sprintf(equality));
        }
        if (i === l){
            throw new Error("Expecting '=' or '<=' or '>=' in equality: %s".sprintf(equality));   
        }
    }
    if (!allowedAttributes.has(token)){
        throw new Error("Attribute name '%s' not valid in equality: %s".sprintf(token, equality));
    }
    constraint.firstAttribute = token;
    constraint.relation = 'equal';
    if (eq[i] == '<'){
        constraint.relation = 'lessThanOrEqual';
        ++i;
        if (i == l || eq[i] != '='){
            throw new Error("Expecting '<=' in equality: %s".sprintf(equality));
        }
    }else if (eq[i] == '>'){
        constraint.relation = 'greaterThanOrEqual';
        ++i;
        if (i == l || eq[i] != '='){
            throw new Error("Expecting '>=' in equality: %s".sprintf(equality));
        }
    }
    i += 1;
    if (i == l){
        throw new Error("Expecting right hand side of equality: %s".sprintf(equality));
    }
    token = '';
    while (i < l && eq[i] != '.' && eq[i] != '*' && eq[i] != '+' && eq[i] != '-' && eq[i] != '@'){
        token += eq[i];
        ++i;
    }
    if (token.length === 0){
        throw new Error("Expecting item or constant on right hand side of equality: %s".sprintf(equality));
    }
    if (i == l){
        constraint.constant = resolveNumber(token);
    }else{
        if (eq[i] == '.'){
            ++i;
            if (!(token in references)){
                constraint.secondItemName = token;
            }else{
                constraint.secondItem = references[token];
            }
            token = '';
            while (i < l && eq[i] != '.' && eq[i] != '*' && eq[i] != '+' && eq[i] != '-' && eq[i] != '@'){
                token += eq[i];
                ++i;
            }
            if (token.length === 0){
                throw new Error("Expecting item attribute right hand side of equality: %s".sprintf(equality));
            }
            if (eq[i] == '.'){
                if (!allowedProperties.has(token)){
                    if (!allowedAttributes.has(token)){
                        throw new Error("Property name '%s' not valid in equality: %s".sprintf(token, equality));
                    }
                }
                constraint.secondProperty = token;
                i += 1;
                token = '';
                while (i < l && eq[i] != '*' && eq[i] != '+' && eq[i] != '-' && eq[i] != '@'){
                    token += eq[i];
                    ++i;
                }
                if (token.length === 0){
                    throw new Error("Expecting item attribute right hand side of equality: %s".sprintf(equality));
                }
            }
            if (!allowedAttributes.has(token)){
                throw new Error("Attribute name '%s' not valid in equality: %s".sprintf(token, equality));
            }
            constraint.secondAttribute = token;
        }else if (eq[i] == '@'){
            constraint.constant = resolveNumber(token);
        }else{
            throw new Error("Unexpected token in equality: %s".sprintf(equality));
        }
        if (i < l && eq[i] == '*'){
            ++i;
            token = '';
            while (i < l && eq[i] != '+' && eq[i] != '-' && eq[i] != '@'){
                token += eq[i];
                ++i;
            }
            if (token.length === 0){
                throw new Error("Expecting multiplier in equality: %s".sprintf(equality));
            }
            constraint.multiplier = resolveNumber(token);
        }
        if (i < l && (eq[i] == '+') || (eq[i] == '-')){
            let sign = eq[i] == '-' ? -1 : 1;
            ++i;
            token = '';
            while (i < l && eq[i] != '@'){
                token += eq[i];
                ++i;
            }
            if (token.length === 0){
                throw new Error("Expecting constant in equality: %s".sprintf(equality));
            }
            constraint.constant = sign * resolveNumber(token);
        }
        if (i < l && eq[i] == '@'){
            ++i;
            token = eq.substr(i);
            if (token.length === 0){
                throw new Error("Expecting priority in equality: %s".sprintf(equality));
            }
            constraint.priority = resolveNumber(token);
        }
    }
    return constraint;
};

var allowedProperties = new Set(['margins']);
var allowedAttributes = new Set(['left', 'right', 'top', 'bottom', 'leading', 'trailing', 'width', 'height', 'centerX', 'centerY']);